packages/compass-components/src/components/compass-components-provider.tsx

```diff
@@ -0,0 +1,132 @@
+
+const darkModeMediaQuery = (() => {
+  return typeof window !== 'undefined' && window.matchMedia
+    ? window.matchMedia('(prefers-color-scheme: dark)')
+    : // A fallback for test environment. This API is supported in browsers
+      // since forever, so there is no way this will be missing in a real
+      // browser / electron env
+      {
+        matches: false,
+        addEventListener() {
+          // noop
+        },
+        removeEventListener() {
+          // noop
+        },
+      };
+})();
+
+function useDarkMode(_darkMode?: boolean) {
+  const isDarkModeControlled = typeof _darkMode !== 'undefined';
+  const [darkMode, setDarkMode] = useState(
+    _darkMode ?? darkModeMediaQuery.matches
+  );
+  useEffect(() => {
+    const onDarkModeMediaQueryChange = () => {
+      setDarkMode(darkModeMediaQuery.matches);
+    };
+    darkModeMediaQuery.addEventListener('change', onDarkModeMediaQueryChange);
+    return () => {
+      darkModeMediaQuery.removeEventListener(
+        'change',
+        onDarkModeMediaQueryChange
+      );
+    };
+  }, []);
+  return isDarkModeControlled ? _darkMode : darkMode;
+}
+
+/**
+ * All compass components providers combined into one for convenience
+ */
+export const CompassComponentsProvider = ({
+  darkMode: _darkMode,
+  children,
+  onNextGuideGue,
+  onNextGuideCueGroup,
+  ...signalHooksProviderProps
+}: CompassComponentsProviderProps) => {
+  const darkMode = useDarkMode(_darkMode);
+
+  // Leafygreen doesn't accept refs for dom elements so we are required to
+  // trigger a state update when the popover ref is resolved. This is very
+  // pricey operation as this re-renders the whole application tree, but there
+  // is literally no way around it with how leafygreen popover works and lucky
+  // for us, this will usually cause a state update only once
+  const [portalContainer, setPortalContainer] = useState<HTMLElement | null>(
+    null
+  );
+  const [scrollContainer, setScrollContainer] = useState<HTMLElement | null>(
+    null
+  );
+
+  const popoverPortalContainer = useMemo(() => {
+    return { portalContainer, scrollContainer };
+  }, [portalContainer, scrollContainer]);
+
+  return (
+    <LeafyGreenProvider
+      darkMode={darkMode}
+      popoverPortalContainer={popoverPortalContainer}
+    >
+      <GuideCueProvider
+        onNext={onNextGuideGue}
+        onNextGroup={onNextGuideCueGroup}
+      >
+        <SignalHooksProvider {...signalHooksProviderProps}>
+          <ConfirmationModalArea>
+            <ToastArea>
+              {typeof children === 'function'
+                ? children({
+                    darkMode,
+                    portalContainerRef: setPortalContainer,
+                    scrollContainerRef: setScrollContainer,
+                  })
+                : children}
+            </ToastArea>
+          </ConfirmationModalArea>
+        </SignalHooksProvider>
+      </GuideCueProvider>
+    </LeafyGreenProvider>
```

Comment:
Another very optional suggestion, maybe there could be a comment describing a rule of thumb regarding which providers go into this component and which don’t? For example, my gut feeling is that we’d always want to keep the service providers for plugins separate, but that might be wrong and different people might have different gut feelings :)

Reply:
Totally, let me extend the current component description

Reply:
Updated the description in https://github.com/mongodb-js/compass/pull/5257/commits/01381da15275b76366aed2e177f6570adee69bb3, how does it look?